numpy/core/src/umath/ufunc_object.c

```diff
@@ -1270,6 +1435,119 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
     for (i = 0; i < nop; i++) {
         Py_XDECREF(out_op[i]);
     }
+    for (i = 0; i < nop; i++) {
+        Py_XSETREF(specified_types[i], NULL);
+    }
+    return -1;
+}
+
+
+/*
+ * This function extracts the datatypes from all arrays.
+ * It fills dtypes with new data.
+  * When an error occurs it return -1 and dtypes is guaranteed
+ * to be NULL initialized.
+ */
+static int
+get_op_dtypes(int nop, int nin, PyArrayObject **arrs, PyArray_Descr **dtypes)
```

Comment:
So this is the logic to find the minimal scalar type. The old is very much the same, but in the `DefaultTypeResolver`, `ResulType` will never be called, instead. If `use_min_scalar` is True, the first loop which `CanCastArrayTo` matches is used (i.e. value based casting is applied during loop search).

Reply:
@eric-wieser just for you to note, I think I decided to try the int7 idea... It should be fully compatible with the old logic, and I suppose while annoying probably fairly straight forward. That also should make this fairly reasonable to move forward at some point.

Reply:
OK, so writing it down also for myself: After thinking about it a bit more seriously, there is a problem with that approach for user dtypes. Right now, if I have a user dtype it will be valid, but rational cannot cast "int7"->rational (since it, as of now) doesn't know about its existence. Of course this is all rather theoretical, but it does show that it is maybe not be a good approach.